### PR TITLE
Add BuildResult log filtering

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

**Comment by henryju** [comment]
Each scanner (Maven, Gradle, MSBuild, CLI) has its own way to print logs. This change looks very specific to me, will probably not work with all scanners, and I don't see why it should be done in Orchestrator and can't be done in calling code. Are you trying to fix an issue (like OOM because of too much logs)?

**Comment by m-g-sonar** [comment]
@henryju Ok, I was also thinking it was not going to work for all the scanners.
The point is that the past few weeks, Languages Team has been bitten a numerous of time by deprecated SonarQube properties suddenly dropped. As we are not monitoring the logs of scans, we missed systematically the associated warnings, usually present prior to the drop.

The idea here would be to monitor the analysis warnings, and add an assertions in each of our tests, so we can start failing when new deprecation arise and do something in advance, rather than only reacting when dropped.

**Comment by henryju** [comment]
The most reliable solution IMO would be to introduce a special mode in scanner (scanner-api for example) that would fail fast on warning. `-Dsonar.scanner.failOnWarn=true`

Another option would be to collect all logs of ITs, and do some extraction.

**Comment by m-g-sonar** [comment]
> Another option would be to collect all logs of ITs, and do some extraction.

That's what I was trying to do here.

> The most reliable solution IMO would be to introduce a special mode in scanner (scanner-api for example) that would fail fast on warning. -Dsonar.scanner.failOnWarn=true

Well, that would not cover my use cases. Our analyzers can perfectly emit warnings, and we want (should? may?) to test that in ITs. The point is that we would like to be able to detect new Scanner/SonarQube warnings in order to have "clean" ITs in term of API as much as possible. Today, this is pretty hard to guarantee.

**Comment by henryju** [comment]
You can keep the getLogLines(Predicate) in Orchestrator if you like, it is generic enough to me. The method that filter based on "WARN" looks too specific to my taste.

**Comment by simonbrandhof** [comment]
Coming back to the proposal of `-Dsonar.scanner.failOnWarn=true`, could you have a test that is configured in a way that all known warnings are dropped? This case would fail each time a new warning is detected.

**Comment by simonbrandhof** [comment]
Otherwise +1 to merge the PR without the method about `WARN`.

**Comment by m-g-sonar** [comment]
> could you have a test that is configured in a way that all known warnings are dropped? This case would fail each time a new warning is detected.

@simonbrandhof yes, and it would work nicely for a large majority of our tests, I guess. At least it would allow us to keep track of new warnings. The one we want to be sure to have, we will test them manually.

@henryju I updated the PR keeping only the method having a `Predicate` as parameter.

File: sonar-orchestrator/src/test/java/com/sonar/orchestrator/build/BuildResultTest.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    Writer writer = buildResult.getLogsWriter();
+    writer.append("Hello,\n");
+    writer.append("World!\n");
+    writer.append("This is a pretty good logging feature.\n");
+    writer.append("Goodbye now.\n");
+
+    assertThat(buildResult.getLogsLines(s -> true)).hasSize(4);
+    assertThat(buildResult.getLogsLines(s -> false)).isEmpty();
+    assertThat(buildResult.getLogsLines(s -> s.contains("Hello") || s.contains("now")))
+      .hasSize(2)
```
Comment (simonbrandhof) [inline sonar-orchestrator/src/test/java/com/sonar/orchestrator/build/BuildResultTest.java]: Just FYI, no need to fix, I think that calling `hasSize()` is useless thanks to the next `containsExactly()`.

**Comment by simonbrandhof** [comment]
Commit renamed with prefix "ORCH-435". I will merge the PR once the checks pass. Thanks @m-g-sonar for the contribution.

**Pull request merged**